php/Terminus/Tunnels.php

```diff
@@ -0,0 +1,182 @@
+
+  /**
+   * Returns an array of open tunnels.
+   */
+  public function get_all() {
+    $cmd = 'ps -fU $USER |grep "ssh -[f]" | awk \'{print $2, $5, $12, $15}\'';
+    $status = $this->shellout($cmd, $lines);
+    if ($status == 0) {
+      $tunnels = array();
+      $sites = new SiteFactory();
+      $keys = array('raw', 'pid', 'created', 'port', 'host', 'type', 'environment', 'site_uuid');
+      // Tunnels are keyed by port, ports are keyed by env.site for mapping.
+      foreach ($lines as $line) {
+        $matches = array();
+        if (preg_match('/([0-9]{1,8}) ([0-9]{1,2}:[0-9]{2}PM|AM) ([0-9]{1,7}):127.0.0.1:[0-9]{1,7} [A-z0-9_]{3,16}\.[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}@(([A-z0-9_]{3,16})\.([A-z0-9_]{3,16})\.([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12})\.drush.in)/', $line, $matches)) {
+          $tunnel = array_combine($keys, $matches);
+          if (isset($data['site_uuid'])) {
+            $site = $sites->getSiteByUUID($tunnel['site_uuid']);
+          }
+          unset($tunnel['raw']);
+          $tunnels[] = $tunnel;
+        }
+      }
+      return($tunnels);
+    }
+    return false;
+  }
+
+  /**
+   * Opens a tunnel if one is not already open.
+   * @param $site (site object) requred: the target site
+   * @param $env (string) required: the target environment
+   * @param $type (string) which binding to connect to
+   * @param $port (int) local port to bind
+   * @param $strict (bool) check hostnames?
+   */
+  public function create($site, $env, $type = 'dbserver', $port = null, $strict = false) {
+    $binding = $site->find_binding($env, $type);
+
+    if (!$binding) {
+      \Terminus::error(sprintf('Could not find %s binding in %s', $type, $env));
+      return false;
+    }
+
+    $tunnel = null;
+    $port = isset($port) ? $port : $binding->port;
+    $rhost = $this->remote_hostname($site->id, $env, $type);
+    if ($tunnel = $this->get_by_port($port)) {
+      if ($rhost == $tunnel['host']) {
+        \Terminus::line(sprintf('Tunnel already open to %s:%d.', $tunnel['host'], $tunnel['port']));
+      }
+      else {
+        \Terminus::error(sprintf('Local port %d is already in use with a different service or host.', $port));
+        return false;
+      }
+    }
+    elseif ($tunnel = $this->get_by_site($site, $env, $type)) {
+      \Terminus::line(sprintf('Tunnel already open to %s:%d.', $tunnel['host'], $tunnel['port']));
+    }
+    else {
+      $cmd = "ssh -f -N -L {$port}:127.0.0.1:{$binding->port} -p 2222 {$env}.{$site->id}@{$rhost}";
+      if (!$strict) {
+        // Keep it quiet
+        $cmd .= ' -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no';
+      }
+      $status = $this->proc_open($cmd);
+
+      if ($status == 0 && $tunnel = $this->get_by_site($site, $env, $type)) {
+        \Terminus::log(sprintf('Tunnel opened to %s:%s.', $tunnel['host'], $tunnel['port']));
+      }
+      else {
+        \Terminus::error('Error setting up the tunnel');
```

Comment:
I can't seem to trigger the success message, even when forwarding is established successfully.
![image](https://cloud.githubusercontent.com/assets/868204/8656145/bcd8a182-294a-11e5-813c-38240593b654.png)